### ISSUE TITLE
PyTorch Version

### ISSUE THREAD

**Issue by Hazqeel09** [HamadYA/GhostFaceNets#27]
I'm trying to convert this model into [PyTorch version](https://github.com/Hazqeel09/GhostFaceNet)

Based on the code here, I understand the flow of Modified GDC as:
DW Conv -> BN -> Dropout -> Conv -> Flatten -> BN -> Activation Func -> Output
<img width="978" alt="Screenshot 2023-10-09 at 10 43 42 AM" src="https://github.com/HamadYA/GhostFaceNets/assets/33782405/812a7731-0684-4cff-8857-02f82a74e690">
But from the paper, before the output, there should be linear layer

Can you check if my implementation [here](https://github.com/Hazqeel09/GhostFaceNet/blob/main/GhostFaceNetV2.py#L60) in PyTorch is correct? Based on my understanding, the model will create embeddings for facial recognition.

**Comment by HamadYA** [comment]
> I'm trying to convert this model into [PyTorch version](https://github.com/Hazqeel09/GhostFaceNet)
> 
> Based on the code here, I understand the flow of Modified GDC as: DW Conv -> BN -> Dropout -> Conv -> Flatten -> BN -> Activation Func -> Output <img alt="Screenshot 2023-10-09 at 10 43 42 AM" width="978" src="https://user-images.githubusercontent.com/33782405/273489368-812a7731-0684-4cff-8857-02f82a74e690.png"> But from the paper, before the output, there should be linear layer
> 
> Can you check if my implementation [here](https://github.com/Hazqeel09/GhostFaceNet/blob/main/GhostFaceNetV2.py#L60) in PyTorch is correct? Based on my understanding, the model will create embeddings for facial recognition.

Hi,
Amazing work!

I am not expert in PyTorch, but I do not think line 76 is correct:

 self.linear = nn.Linear(flattened_features, num_classes) if num_classes else nn.Identity()

The flow of Modified GDC is correct
DW Conv -> BN -> Dropout -> Conv -> Flatten -> BN -> Activation Func -> Output
The final activation function is simply a linear activation function that leads to my 512-d output. I think you can simply use:

x = x.view(x.size(0), -1)

or
import torch.nn as nn

embedding = nn.BatchNorm1d(emb_shape, momentum=bn_momentum, eps=bn_epsilon, affine=scale)

embedding_fp32 = nn.Identity()(embedding)


And you can optionally use:

#optional
embedding_fp32 = embedding.to(torch.float32)

**Comment by Hazqeel09** [comment]
I see, I misunderstood about the 512 embeddings. Will fix it, thank you.

**Comment by HamadYA** [comment]
> activation function is simply a linear activation function that leads to my 512-d out

Thanks for your amazing work!
I am very glad that someone is actually implementing the model using PyTorch.
Keep up the good work!

**Comment by Hazqeel09** [comment]
Thank you for your work too!
